### DISCUSSION TITLE
Serial to TCP Bridge BaudRate more than 115200

### DISCUSSION THREAD

**Discussion by shantur** [arendst/Tasmota#21523] [Support request]
Hi all,

I am trying to get a faster connection for the Serial to TCP bridge. Currently the max baudrate is limited to 115200 how ever the same for Serial and SoftwareSerial is 19,660,500.

Is is some hardware limitation or can TCP Bridge baud rate be increased?

Thanks

**Comment by sfromis** [comment]
As mentioned on Discord:
> Not sure where the hardware max is, but I'd say that software serial cannot even get in the ballpark of 19660500 bps.
Still, I'd expect 115200 not to be any "hard limit", you could try changing https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_41_tcp_bridge.ino#L215

**Comment by Jason2866** [comment]
From datasheet 5Mbs are max for UART. Not saying this is possible with Tasmota.

**Reply by shantur** [reply]
As a PR, should the check for maximum limit be removed to match Software Serial and Hardware serial or limit it to 5Mbs?

**Reply by Jason2866** [reply]
No, the check should only be removed if it is verified working with a higher speed. Tasmota is not just serving UARTs ;-)

**Reply by sfromis** [reply]
Your are certainly not getting anywhere near 5 mbit with software serial, and I'd not bank on it for hardware serial without careful testing.

That was why I suggested to **try** changing.

**Reply by shantur** [reply]
SoftwareSerial and Hardware serial don't have any upper limits on BaudRate so wasn't sure if we should align them

https://github.com/arendst/Tasmota/blob/b61ea17ce14a7a836973892fc51cef118a454e8b/tasmota/tasmota_xdrv_driver/xdrv_08_serial_bridge.ino#L268

https://github.com/arendst/Tasmota/blob/b61ea17ce14a7a836973892fc51cef118a454e8b/tasmota/tasmota_support/support_command.ino#L1979

**Reply by sfromis** [reply]
I'd say to just forget about using software serial for TCP bridge, even at 115200 which is already not likely to work. And we were **not** talking about "upper limits", but about what would work doing real-life testing.

**Reply by s-hadinger** [reply]
On esp8266, software serial is a challenge at 115200. I had to introduce some special handling to auto-correct wrong messages with Zigbee. See https://tasmota.github.io/docs/Zigbee-Internals/#cc2530-serial-protocol

Hardware serial works more reliably but then the length of wires and electrical noise can also limit speed

**Reply by shantur** [reply]
With a local build removing the upper limit check for ( 115200 ), I was able to go upto 306000 due to a uint8_t settings->tcp_baudrate.
I hacked settings->tcp_baudrate to be a uint16_t and I was be able to test 1.5Mbaud with flashing Zigbee Bridge Pro with llama-bsl over TCP. I wasn't able to test more than 1.5Mbaud as Serial bootloader supports a max 1.5Mbaud.

It seems to me that on ESP32 Tasmota can easily support more than 115200 baud rate over TCP Serial bridge.

I could have done a PR but this would need a settings change and if I remember correctly its a delicate balance which maybe @arendst can manage.

**Reply by shantur** [reply]
Raised a PR to remove limit and match HardwareSerial (CmndBaudRate) and SoftwareSerial (CmndSBaudRate)
https://github.com/arendst/Tasmota/pull/21528

**Reply by sfromis** [reply]
My main concern about TCP bridge and speeds is not so much sending serial (which due to being async is very tolerant about not keeping full speed), but about *receiving* serial data at a high and *sustained* data rate; where it really starts to push things.
